src/connection.c

```diff
@@ -330,3 +328,24 @@ ConnectionGetOptionValue(PGconn *connection, char *optionKeyword)
 
 	return optionValue;
 }
+
+ShardId DoForAllShards(List* shardConnections, ShardAction action, void* arg)
+{
+	ListCell *listCell;
+	int i = 0;
+
+	foreach (listCell, shardConnections)
+	{
+		ShardConnections* shardConn = (ShardConnections*)lfirst(listCell);
+		bool allOk = true;
+		for (i = 0; i < shardConn->replicaCount; i++) 
+		{						
+			allOk &= shardConn->status[i] = action(shardConn->shardId, shardConn->conn[i], arg, shardConn->status[i]);
+		}
+		if (!allOk) 
```

Comment:
This is why "status" variable is used.
Please notice that PgCopyAbortTransaction and PgCopyEndTransaction always return true, so in any case all transactions in all connections will be completed and there will be no unclosed prepared transactions.